ic-factory/src/top_up.rs

```diff
@@ -0,0 +1,148 @@
+use candid::Principal;
+use cycles_minting_canister::{
+    IcpXdrConversionRateCertifiedResponse, NotifyError, NotifyTopUp, DEFAULT_CYCLES_PER_XDR,
+    MEMO_TOP_UP_CANISTER,
+};
+use ic_base_types::{CanisterId, PrincipalId};
+use ic_canister::virtual_canister_call;
+use ledger_canister::{
+    AccountIdentifier, BlockHeight, SendArgs, Subaccount, Tokens, DEFAULT_TRANSFER_FEE,
+    TOKEN_SUBDIVIDABLE_BY,
+};
+use num_traits::ToPrimitive;
+
+use crate::error::FactoryError;
+
+const CYCLE_MINTING_CANISTER: &str = "rkp4c-7iaaa-aaaaa-aaaca-cai";
+
+/// This function calculates the amount required for minting cycles for a canister.
+pub async fn cycles_to_icp(cycles: u64) -> Result<u64, FactoryError> {
+    let rate = get_conversion_rate().await?.data;
+
+    let icp_per_xdr = 10_000.0 / rate.xdr_permyriad_per_icp as f64;
+
+    // Convert cycles to XDRs - 1 XDR = 10^12 cycles
+    let xdr = cycles as f64 / DEFAULT_CYCLES_PER_XDR as f64;
+
+    let icp = xdr * icp_per_xdr;
+
+    (icp * TOKEN_SUBDIVIDABLE_BY as f64)
+        .to_u64()
+        .ok_or_else(|| FactoryError::GenericError("Failed to convert cycles to ICP".to_string()))
+}
+
+async fn get_conversion_rate() -> Result<IcpXdrConversionRateCertifiedResponse, FactoryError> {
+    let principal = Principal::from_text(CYCLE_MINTING_CANISTER).expect("const conversion");
+
+    virtual_canister_call!(
+        principal,
+        "get_icp_xdr_conversion_rate",
+        (),
+        IcpXdrConversionRateCertifiedResponse
+    )
+    .await
+    .map_err(|e| FactoryError::GenericError(e.1))
+}
+
+pub(crate) async fn transfer_icp_to_cmc(
+    amount: u64,
+    ledger: Principal,
+    caller_subaccount: Subaccount,
+) -> Result<BlockHeight, FactoryError> {
+    let canister_minting_principal =
+        Principal::from_text(CYCLE_MINTING_CANISTER).expect("const conversion");
+
+    let canister_id = ic_canister::ic_kit::ic::id();
+    let to = AccountIdentifier::new(
+        canister_minting_principal.into(),
+        Some((&PrincipalId::from(canister_id)).into()),
+    );
+
+    let args = SendArgs {
+        memo: MEMO_TOP_UP_CANISTER,
+        amount: Tokens::from_e8s(amount),
+        fee: DEFAULT_TRANSFER_FEE,
+        from_subaccount: Some(caller_subaccount),
+        to,
+        created_at_time: None,
+    };
+
+    virtual_canister_call!(ledger, "send_dfx", (args,), u64)
```

Comment:
Do not use `send_dfx` method. Use `transfer` instead. `send_dfx` is for internal use and can be changed at any moment.